File: triangle.py
# problem: https://leetcode.com/problems/triangle/
# Runtime: 56 ms, faster than 79.12% of Python3 online submissions for Triangle.
# Memory Usage: 14.9 MB, less than 40.45% of Python3 online submissions for Triangle.

from typing import List


class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        for i in range(1, len(triangle)):
            for j in range(len(triangle[i])):
                if j == 0:
                    triangle[i][j] += triangle[i-1][j]
                elif j == (len(triangle[i]) - 1):
                    triangle[i][j] += triangle[i-1][j-1]
                else:
                    triangle[i][j] += min(triangle[i-1][j], triangle[i-1][j-1])
        return min(triangle[-1])

File: triangle.go
// problem: https://leetcode.com/problems/triangle/
// Runtime: 4 ms, faster than 94.63% of Go online submissions for Triangle.
// Memory Usage: 3.1 MB, less than 100.00% of Go online submissions for Triangle.

package algorithms

func max(x int, y int) int {
	if x > y {
		return y
	}
	return x
}

func minimumTotal(triangle [][]int) int {
	for i := len(triangle) - 2; i >= 0; i-- {
		for j := 0; j < len(triangle[i]); j++ {
			triangle[i][j] += max(triangle[i+1][j], triangle[i+1][j+1])
		}
	}
	return triangle[0][0]
}
